template-compose/data/src/main/java/co/nimblehq/template/compose/data/repository/UserPreferencesRepositoryImpl.kt

```diff
@@ -0,0 +1,54 @@
+package co.nimblehq.template.compose.data.repository
+
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.*
+import co.nimblehq.template.compose.domain.model.UserPreferences
+import co.nimblehq.template.compose.domain.repository.UserPreferencesRepository
+import kotlinx.coroutines.flow.*
+import java.io.IOException
+import javax.inject.Inject
+
+class UserPreferencesRepositoryImpl @Inject constructor(private val dataStore: DataStore<Preferences>) :
+    UserPreferencesRepository {
+
+    private val TAG: String = "UserPreferencesRepo"
+
+    private object PreferencesKeys {
+        val USERNAME = stringPreferencesKey("username")
+        val IS_ACTIVE = booleanPreferencesKey("is_active")
+    }
+
+    override fun getUsernamePreferences(): Flow<UserPreferences> = dataStore.data
+        .catch { exception ->
+            // dataStore.data throws an IOException when an error is encountered when reading data
+            if (exception is IOException) {
+                Log.e(TAG, "Error reading preferences.", exception)
+                emit(emptyPreferences())
+            } else {
+                throw exception
+            }
+        }.map { preferences ->
+            mapUserPreferences(preferences)
+        }
+
+    override suspend fun updateUsernamePreferences(usernamePreferences: UserPreferences) {
+        dataStore.edit { preferences ->
+            val currentUserPreferences = mapUserPreferences(preferences)
+
+            if (usernamePreferences.username != currentUserPreferences.username) {
```

Comment:
Is it necessary to check the current value? I think it would be faster to do it without 💭

Reply:
@Tuubz What do you think? Which one is worthier, checking the value before the update or updating every time calling the function? 🤔

Reply:
@lydiasama With the current approach, we will always need to fetch the values first, which counts as **one** action 1️⃣ 

If we then need to update the value too, that would count as **another** action. 2️⃣ 

Summary:
- **"checking the value before the update"** : we will always have _one_ action, and possibly have _two_ actions.
- **"updating every time"** : we can be sure it will always execute _one_ action

However, it could be that **updating** is more "heavy" compared to **"fetching"** 💭 (not sure)

Please let me know if I got the wrong idea here, what do you think? 😁 

cc @nimblehq/android-chapter

Reply:
@Tuubz What if this case, it updates and gets an error with the same value? Do we need to handle the exception like `Cannot update preferences` like the different value? If we don't check the same value before, perhaps, we need to check when getting the exception to handle in case `Cannot update with new value to preferences` or `Cannot update with the same value -> can ignore this exception`. What do you think?

Reply:
@lydiasama Oh, does datastore throw an error if it updates with the same value? 😯

Reply:
@Tuubz Nooo, I meant if it update with the same error but get the error while updating for some reason.

Reply:
@Tuubz @lydiasama IMO, I think we should work on a `sample` project first and try to integrate, So we can debug more and prove the concept of your idea 🙏